Always parses BBC using the recipent's language in email notifications

Sesquipedalian opened the pull request:
Doing this requires using `parse_bbc()` in the langauge of the recipient of the email notification rather than in the language of the current user.

Main changes:
- `parse_bbc()` can now remember versions of the $bbc_codes array for multiple languages.
- For new post notifications, the body text of the message is now censored and parsed in CreatePost-Notify.php rather than in Subs-Post.php's `sendNotifications()`. This allows us to tailor the output of `parse_bbc()` depending on the recipient's language.
- Similarly, some code has been moved around and modified in the `sendpm()` function of Subs-Post.php.

~TODO: This code temporarily resets `$txt` to a blank slate, populates it with the contents of the index language file for the recipient's language, and then restores `$txt` to its original state when we are done. This is fine for parsing standard BBC, but it will know nothing about `$txt` strings stored in other files, e.g. in those added by mods. Should probably do something about that...~ ← Done.

Fixes #1407 (which dates all the way back to 2006: http://dev.simplemachines.org/mantis/view.php?id=52)

Sesquipedalian committed:
Always parses BBC using the recipent's language in email notifications
Doing this requires using `parse_bbc()` in the langauge of the recipient of the email notification rather than in the language of the current user.

Main changes:
- `parse_bbc()` can now remember versions of the $bbc_codes array for multiple languages.
- For new post notifications, the body text of the message is now censored and parsed in CreatePost-Notify.php rather than in Subs-Post.php's `sendNotifications()`. This allows us to tailor the output of `parse_bbc()` depending on the recipient's language.
- Similarly, some code has been moved around and modified in the `sendpm()` function of Subs-Post.php.

TODO: This code temporarily resets `$txt` to a blank slate, populates it with the contents of the index language file for the recipient's language, and the restores `$txt` to its original state when we are done. This is fine for parsing standard BBC, but it will know nothing about `$txt` strings stored in other files, e.g. in those added by mods. Should probably do something about that...

Fixes #1407 (which dates all the way back to 2006: http://dev.simplemachines.org/mantis/view.php?id=52)

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Adds a $smcFunc['htmlspecialchars']() that I forgot in prev. commit

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

live627 commented:
Unrelated: I don't have access to Mantis. I thought it was closed.

On Thu, Apr 4, 2019 at 9:38 PM Jon Stovell <notifications@github.com> wrote:

> Doing this requires using parse_bbc() in the langauge of the recipient of
> the email notification rather than in the language of the current user.
>
> Main changes:
>
>    - parse_bbc() can now remember versions of the $bbc_codes array for
>    multiple languages.
>    - For new post notifications, the body text of the message is now
>    censored and parsed in CreatePost-Notify.php rather than in Subs-Post.php's
>    sendNotifications(). This allows us to tailor the output of parse_bbc()
>    depending on the recipient's language.
>    - Similarly, some code has been moved around and modified in the
>    sendpm() function of Subs-Post.php.
>
> TODO: This code temporarily resets $txt to a blank slate, populates it
> with the contents of the index language file for the recipient's language,
> and the restores $txt to its original state when we are done. This is
> fine for parsing standard BBC, but it will know nothing about $txt
> strings stored in other files, e.g. in those added by mods. Should probably
> do something about that...
>
> Fixes #1407 <https://github.com/SimpleMachines/SMF2.1/issues/1407> (which
> dates all the way back to 2006:
> http://dev.simplemachines.org/mantis/view.php?id=52)
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/SimpleMachines/SMF2.1/pull/5575
> Commit Summary
>
>    - Always parses BBC using the recipent's language in email
>    notifications
>
> File Changes
>
>    - *M* Sources/Subs-Post.php
>    <https://github.com/SimpleMachines/SMF2.1/pull/5575/files#diff-0> (49)
>    - *M* Sources/Subs.php
>    <https://github.com/SimpleMachines/SMF2.1/pull/5575/files#diff-1> (10)
>    - *M* Sources/tasks/CreatePost-Notify.php
>    <https://github.com/SimpleMachines/SMF2.1/pull/5575/files#diff-2> (33)
>
> Patch Links:
>
>    - https://github.com/SimpleMachines/SMF2.1/pull/5575.patch
>    - https://github.com/SimpleMachines/SMF2.1/pull/5575.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SimpleMachines/SMF2.1/pull/5575>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAaWt3owSCturz3LwEgLnFPnVeyb_yo0ks5vdtM4gaJpZM4ceJd2>
> .
>

Sesquipedalian commented:
Mantis is closed now, yes. That commit message was originally authored... well, a while ago.

live627 commented on Sources/tasks/CreatePost-Notify.php:
```diff
@@ -194,18 +197,38 @@ public function execute()
+			// Censor and parse BBC in the receiver's language. Only do each language once.
+			if (empty($parsed_message[$receiver_lang]))
+			{
+				if ($receiver_lang != $user_info['language'])
```
Err, I don't think that `$user_info` is available in cron.

Sesquipedalian replied:
Derp. Although `$user_info` itself is, `$user_info['language']` isn't set.

Sesquipedalian replied:
Fixed in 3d64b0e

Sesquipedalian committed:
Simplifies language loading in CreatePost-Notify

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Loads Modifications language file in case custom BBC put strings there

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Removes redundant redundancy that was redundant

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Simplifies and improves language loading in sendpm()

Signed-off-by: Jon Stovell <jonstovell@gmail.com>

Sesquipedalian committed:
Simplifies logic regarding $parse_tags and static vars in parse_bbc()
Since $bbc_codes isn't a static variable anymore ($bbc_lang_locales does that job now), there's no need for the convoluted juggling that had to be done before when the $parse_tags parameter wasn't empty.

Signed-off-by: Jon Stovell <jonstovell@gmail.com>
